src/protocol/sort/generate_permutation_opt.rs

```diff
@@ -0,0 +1,197 @@
+///
+/// Steps
+/// For the `num_multi_bits`
+/// 1. Get replicated shares in Field using modulus conversion
+/// 2. Compute bit permutation that sorts 0..`num_multi_bits`
+/// For `num_multi_bits` to N-1th bit of input share
+/// 1. Shuffle and reveal the i-1th composition
+/// 2. Get replicated shares in Field using modulus conversion
+/// 3. Sort i..i+`num_multi_bits` bits based on i-1th bits by applying i-1th composition on all these bits
+/// 4  Compute bit permutation that sorts i..i+`num_multi_bits`
+/// 5. Compute ith composition by composing i-1th composition on ith permutation
+/// In the end, n-1th composition is returned. This is the permutation which sorts the inputs
+///
+/// # Errors
+/// If any underlying protocol fails
+/// # Panics
+/// Panics if input doesn't have same number of bits as `num_bits`
+
+pub async fn generate_permutation_opt<F>(
+    ctx: SemiHonestContext<'_, F>,
+    sort_keys: &[Vec<Replicated<F>>],
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Result<Vec<Replicated<F>>, Error>
+where
+    F: Field,
+{
+    let ctx_0 = ctx.narrow(&Sort(0));
+    assert_eq!(sort_keys.len(), num_bits as usize);
+
+    let last_bit_num = if num_multi_bits > num_bits {
+        num_bits
+    } else {
+        num_multi_bits
+    };
+
+    let lsb_permutation = multi_bit_permutation(
+        ctx_0.narrow(&BitPermutationStep),
+        &sort_keys[0..last_bit_num.try_into().unwrap()],
+    )
+    .await?;
+
+    let input_len = u32::try_from(sort_keys[0].len()).unwrap(); // safe, we don't sort more that 1B rows
+
+    let mut composed_less_significant_bits_permutation = lsb_permutation;
+    for bit_num in (num_multi_bits..num_bits).step_by(num_multi_bits.try_into().unwrap()) {
+        let ctx_bit = ctx.narrow(&Sort(bit_num));
+        let revealed_and_random_permutations = shuffle_and_reveal_permutation(
+            ctx_bit.narrow(&ShuffleRevealPermutation),
+            input_len,
+            composed_less_significant_bits_permutation,
+        )
+        .await?;
+
+        let (randoms_for_shuffle0, randoms_for_shuffle1, revealed) = (
+            revealed_and_random_permutations
+                .randoms_for_shuffle
+                .0
+                .as_slice(),
+            revealed_and_random_permutations
+                .randoms_for_shuffle
+                .1
+                .as_slice(),
+            revealed_and_random_permutations.revealed.as_slice(),
+        );
+
+        let last_bit_num = if bit_num + num_multi_bits > num_bits {
+            num_bits
+        } else {
+            bit_num + num_multi_bits
+        };
+
+        let futures =
+            (bit_num..last_bit_num)
+                .zip(repeat(ctx_bit.clone()))
+                .map(|(idx, ctx_bit)| async move {
+                    secureapplyinv(
+                        ctx_bit.narrow(&MultiApplyInv(idx)),
+                        sort_keys[idx as usize].clone(),
```

Comment:
Hmmm... This isn't what I had expected to see.

You're calling `secureapplyinv` in a loop. Can't we just apply this one time, to all of the affected rows? The same way that we eventually apply the sort permutation to ALL of the rows of the inputs to the IPA query.

Reply:
Each loop iteration each async move so should happen in parallel.

Reply:
Sure, it's parallelized, but is there any more communication going on this way?